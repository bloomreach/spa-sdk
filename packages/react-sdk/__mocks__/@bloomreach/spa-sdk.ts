/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const spaSdk = jest.requireActual('@bloomreach/spa-sdk');

export const META_POSITION_BEGIN = spaSdk.META_POSITION_BEGIN;
export const META_POSITION_END = spaSdk.META_POSITION_END;

export function mockMeta(data: string, position: string) {
  return {
    getData: jest.fn().mockReturnValue(data),
    getPosition: jest.fn().mockReturnValue(position),
  };
}

export function mockNoCommentMeta() {
  return mockMeta('not-a-comment', META_POSITION_BEGIN);
}

const componentMock = {
  getMeta: jest.fn().mockReturnValue([
    mockMeta('meta-begin', META_POSITION_BEGIN),
    mockMeta('meta-end', META_POSITION_END),
  ]),
};

export const pageMock = {
  getComponent: jest.fn().mockReturnValue(componentMock),
  getContent: jest.fn(),
  getTitle: jest.fn(),
  sync: jest.fn(),
};

export const initialize = jest.fn();
export const destroy = jest.fn();

export function isMetaComment(value: any) {
  return value.getData() !== 'not-a-comment';
}
