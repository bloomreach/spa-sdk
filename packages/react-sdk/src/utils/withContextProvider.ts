// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';
import React, { ReactElement } from 'react';

export const withContextProvider = (context?: Record<string, any>, children?: ReactElement): React.ReactElement => {
  const childContextTypes = (): Record<string, string> | undefined => context
    && Object.keys(context).reduce((obj: Record<string, any>, key: string) => {
      obj[key] = PropTypes.any;
      return obj;
    }, {});

  class ContextProvider extends React.Component {
    getChildContext(): Record<string, string> | undefined {
      return context;
    }

    render(): React.ReactNode {
      return children;
    }
  }

  (ContextProvider as any).displayName = 'ContextProvider';
  (ContextProvider as any).childContextTypes = childContextTypes();

  return React.createElement(ContextProvider, null);
};
