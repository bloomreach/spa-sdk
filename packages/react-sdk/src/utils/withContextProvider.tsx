/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement } from 'react';
import { BrPageContext } from '../page';
import { BrComponentContext, BrMappingContext } from '../component';

export const withPageContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrPageContext.Provider value={context}>{children}</BrPageContext.Provider>;
};

export const withMappingContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return (
    <BrMappingContext.Provider value={context}>
      {children}
    </BrMappingContext.Provider>
  );
};

export const withComponentContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrComponentContext.Provider value={context}>{children}</BrComponentContext.Provider>;
};
