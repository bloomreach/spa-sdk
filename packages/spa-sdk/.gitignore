/coverage
/docs
/lib
/node_modules
.DS_store