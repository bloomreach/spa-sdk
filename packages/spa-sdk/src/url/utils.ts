/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

interface Url {
  hash: string;
  origin: string;
  pathname: string;
  search: string;
  searchParams: URLSearchParams;
  path: string;
}

export function buildUrl(url: Partial<Url>) {
  const searchParams = url.searchParams?.toString() ?? '';
  const search = url.search ?? `${searchParams && `?${searchParams}`}`;
  const path = url.path ?? `${url.pathname ?? ''}${search}${url.hash ?? ''}`;

  return `${url.origin ?? ''}${path}`;
}

function isMatchedOrigin(origin: string, baseOrigin: string) {
  return !baseOrigin || !origin || baseOrigin === origin;
}

function isMatchedPathname(pathname: string, basePathname: string) {
  return !basePathname || pathname.startsWith(basePathname);
}

function isMatchedQuery(search: URLSearchParams, baseSearch: URLSearchParams) {
  let match = true;
  baseSearch.forEach((value, key) => {
    match = match && (!value && search.has(key) || search.getAll(key).includes(value));
  });

  return match;
}

export function isMatched(link: string, base = '') {
  const linkUrl = parseUrl(link);
  const baseUrl = parseUrl(base);

  return isMatchedOrigin(linkUrl.origin, baseUrl.origin)
    && isMatchedPathname(linkUrl.pathname, baseUrl.pathname)
    && isMatchedQuery(linkUrl.searchParams, baseUrl.searchParams);
}

export function mergeSearchParams(params: URLSearchParams, ...rest: URLSearchParams[]) {
  const result = new URLSearchParams(params);
  rest.forEach(params => params.forEach((value, key) => result.set(key, value)));

  return result;
}

export function parseUrl(url: string): Url {
  // URL constructor requires either a valid URL or a base URL.
  // Since this function returns a pathname, we can safely pass a fake host to be able to resolve relative URLs.
  const parsedUrl = url ? new URL(url, 'http://example.com') : {} as URL;
  const { hash = '', search = '', searchParams = new URLSearchParams() } = parsedUrl;

  // For links like `//example.com?query#hash` pathname will be `/` so we need to strip query and hash parameters first.
  let origin = url.substring(0, url.length - search.length - hash.length);
  let { pathname = '' } = parsedUrl;
  if (!origin.endsWith(pathname)) {
    pathname = pathname.substring(1);
  }

  origin = origin.substring(0, origin.length - pathname.length);

  return { hash, origin, pathname, search, searchParams, path: `${pathname}${search}${hash}` };
}
