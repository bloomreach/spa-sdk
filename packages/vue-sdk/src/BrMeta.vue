<!--
  Copyright 2023-2025 Bloomreach

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    https://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <span v-if="meta && meta.length > 0" style="display: none;" ref="head"/>
  <slot/>
  <span v-if="meta && meta.length > 0" style="display: none;" ref="tail"/>
</template>

<script setup lang="ts">
import type { MetaCollection } from '@bloomreach/spa-sdk';
import { getCurrentInstance, onBeforeUnmount, onBeforeUpdate, onMounted, onUpdated, ref, watch } from 'vue';

const props = defineProps<{ meta?: MetaCollection }>();
const meta = ref(props.meta);
let clear: ReturnType<MetaCollection['render']> | undefined;

const head = ref<HTMLSpanElement>();
const tail = ref<HTMLSpanElement>();

const inject = () => {
  if (!head.value?.nextSibling || !tail.value) {
    return;
  }

  clear = meta.value?.render(head.value.nextSibling, tail.value);
};

watch(meta, () => getCurrentInstance()?.proxy?.$forceUpdate(), { deep: false });
onMounted(() => inject());
onUpdated(() => inject());
onBeforeUnmount(() => clear?.());
onBeforeUpdate(() => clear?.());
</script>
