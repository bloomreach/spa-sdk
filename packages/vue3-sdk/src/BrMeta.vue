<!--
  - Copyright 2023 Bloomreach
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -   http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <slot></slot>
</template>

<script setup lang="ts">
import type { MetaCollection } from '@bloomreach/spa-sdk';
import {
  defineProps,
  getCurrentInstance,
  onBeforeUnmount,
  onBeforeUpdate,
  onMounted,
  onUpdated,
  ref,
  watch,
} from 'vue';

const { meta } = defineProps<{ meta?: MetaCollection }>();
const metaRef = ref(meta);
let clear: ReturnType<MetaCollection['render']> | undefined;
const inject = () => {
  const { proxy } = getCurrentInstance()!;
  const el = proxy?.$el as Node;
  clear = el && meta?.render(el, el);
};
watch(metaRef, () => {
  const { proxy } = getCurrentInstance()!;
  proxy?.$forceUpdate();
}, { deep: false });

onMounted(() => inject());
onBeforeUpdate(() => clear?.());
onUpdated(() => inject());
onBeforeUnmount(() => clear?.());
</script>
