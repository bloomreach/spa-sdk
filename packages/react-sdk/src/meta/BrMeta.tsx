/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { MetaComment, Meta, META_POSITION_BEGIN, META_POSITION_END, isMetaComment } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: Meta[];
}

export class BrMeta extends React.Component<BrMetaProps> {
  private comments: Comment[] = [];
  private head?: Element | Text;
  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.renderMeta();
  }

  componentDidUpdate() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.removeMeta();
    this.renderMeta();
  }

  componentWillUnmount() {
    this.removeMeta();
  }

  private removeMeta() {
    this.comments.splice(0).forEach(comment => comment.remove());
  }

  private renderMeta() {
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_BEGIN)
      .forEach(this.renderMetaComment.bind(this));
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_END)
      .reverse()
      .forEach(this.renderMetaComment.bind(this));
  }

  private renderMetaComment(meta: MetaComment) {
    const comment = this.head!.ownerDocument!.createComment(meta.getData());
    this.comments.push(comment);

    if (meta.getPosition() === META_POSITION_BEGIN) {
      return void this.head!.parentNode!.insertBefore(comment, this.head!);
    }

    if (!this.tailRef.current!.nextSibling) {
      return void this.tailRef.current!.parentNode!.appendChild(comment);
    }

    this.tailRef.current!.parentNode!.insertBefore(comment, this.tailRef.current!.nextSibling);
  }

  render() {
    return (
      <>
        {this.props.children}
        {this.props.meta.length ? <span style={{ display: 'none' }} ref={this.tailRef} /> : null}
      </>
    );
  }
}
