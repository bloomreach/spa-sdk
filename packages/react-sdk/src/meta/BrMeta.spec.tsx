/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mocked } from 'ts-jest/utils';
import { mount } from 'enzyme';
import { MetaComment, META_POSITION_BEGIN, META_POSITION_END, isMetaComment } from '@bloomreach/spa-sdk';
import { BrMeta } from './BrMeta';

describe('BrMeta', () => {
  const meta = [
    new class implements MetaComment {
      getData = jest.fn(() => 'begin comment 1');
      getPosition = jest.fn(() => META_POSITION_BEGIN as any);
    },
    new class implements MetaComment {
      getData = jest.fn(() => 'begin comment 2');
      getPosition = jest.fn(() => META_POSITION_BEGIN as any);
    },
    new class implements MetaComment {
      getData = jest.fn(() => 'end comment 1');
      getPosition = jest.fn(() => META_POSITION_END as any);
    },
    new class implements MetaComment {
      getData = jest.fn(() => 'end comment 2');
      getPosition = jest.fn(() => META_POSITION_END as any);
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('componentDidMount', () => {
    it('should render comments surrounding children', () => {
      mocked(isMetaComment).mockReturnValue(true);

      const wrapper = mount((
        <div>
          <BrMeta meta={meta}>
            <a/>
            <b/>
          </BrMeta>
        </div>
      ));

      mocked(isMetaComment).mockReset();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render comment meta only', () => {
      mocked(isMetaComment).mockReturnValueOnce(true);

      const wrapper = mount((
        <div>
          <BrMeta meta={meta}/>
        </div>
      ));

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('componentDidUpdate', () => {
    it('should rerender comments on update', () => {
      mocked(isMetaComment).mockReturnValue(true);

      const container = document.createElement('div');
      const wrapper = mount(<BrMeta meta={meta}><a/></BrMeta>, { attachTo: container });
      wrapper.setProps({ meta: [meta[0], meta[2]] });

      mocked(isMetaComment).mockReset();

      expect(container.innerHTML).toMatchSnapshot();
    });
  });

  describe('componentWillUnmount', () => {
    it('should remove the comment when the component unmounts', () => {
      mocked(isMetaComment).mockReturnValue(true);

      const container = document.createElement('div');
      const wrapper = mount(<div><BrMeta meta={meta} /></div>, { attachTo: container });
      wrapper.detach();

      expect(container.innerHTML).toBe('');
    });
  });

  describe('render', () => {
    it('should render only children if there is no meta', () => {
      const wrapper = mount((
        <div>
          <BrMeta meta={[]}>
            <a/>
            <b/>
          </BrMeta>
        </div>
      ));

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
