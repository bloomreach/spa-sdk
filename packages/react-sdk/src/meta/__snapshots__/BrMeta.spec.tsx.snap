// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrMeta componentDidMount should render comment meta only 1`] = `"<div><!--begin comment 1--><span style=\\"display: none;\\"></span></div>"`;

exports[`BrMeta componentDidMount should render comments surrounding children 1`] = `"<div><!--begin comment 1--><!--begin comment 2--><a></a><b></b><span style=\\"display: none;\\"></span><!--end comment 1--><!--end comment 2--></div>"`;

exports[`BrMeta componentDidUpdate should rerender comments on update 1`] = `"<!--begin comment 1--><a></a><span style=\\"display: none;\\"></span><!--end comment 1-->"`;

exports[`BrMeta render should render only children if there is no meta 1`] = `"<div><a></a><b></b></div>"`;
